File: Chapter3/Invpend_solver.m
%
%   Simulation of Inverted Pendulum
%   Copyright Hamid D. Taghirad 2013
%
clear all
clf, echo on
tspan=[0 1];
x0=[0;0.1;0;0];

[t,x] = ode45(@inverted_pendulum,tspan, x0, ...
    odeset('OutputFcn','odeplot','MaxStep',1e-2)); 

plot(t,x(:,1),'k',t,x(:,2),'-.k'),grid
xlabel('Time (sec)')
ylabel('State Variables')
legend('x (m)', '\theta (rad)')
set(findall(figure(1),'type','line'),'linewidth',2)

File: Chapter3/Invpend_solver.py
# Import necessary libraries
import numpy as np
from scipy.integrate import odeint
import matplotlib.pyplot as plt

# Define the model of the inverted pendulum
def inverted_pendulum(x, t):
    g = 9.8
    l = 1.0
    m = 1.0
    M = 1.0
    
    d1 = M + m * (1 - np.cos(x[1])**2)
    d2 = l * d1
    
    F = 0  # No input
    
    xp = np.zeros(4)
    xp[0] = x[2]
    xp[1] = x[3]
    xp[2] = (F + m * l * x[3]**2 * np.sin(x[1]) - m * g * np.sin(x[1]) * np.cos(x[1])) / d1
    xp[3] = (-F * np.cos(x[1]) - m * l * x[3]**2 * np.sin(x[1]) * np.cos(x[1]) + (M + m) * g * np.sin(x[1])) / d2
    
    return xp

# Initial conditions: [x, theta, v, omega]
x0 = [0, 0.1, 0, 0]  # Initial state

# Time vector for the simulation
tspan = np.linspace(0, 1, 100)  # From 0 to 1 second, 100 points

# Solve the system of ODEs
x = odeint(inverted_pendulum, x0, tspan)

# Plot state variables
plt.figure(figsize=(10, 6))
plt.plot(tspan, x[:, 0], 'k', linewidth=2, label='x (m)')
plt.plot(tspan, x[:, 1], '-.k', linewidth=2, label=r'$\theta$ (rad)')
plt.grid(True)
plt.xlabel('Time (sec)')
plt.ylabel('State Variables')
plt.legend()
plt.title('Inverted Pendulum Simulation')
plt.show()
